app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb

```diff
@@ -0,0 +1,134 @@
+      vm_name           = clone_task_ref[3]
+      vm_resource_group = clone_task_ref[1]
+      instance          = vms.get(vm_name, vm_resource_group)
+      status            = instance.properties.provisioning_state
+
+      return true if status == "Succeeded"
+      return false, status
+    end
+  end
+
+  def find_destination_in_vmdb(vm_uid)
+    ems_ref = vm_uid.join("\\")
+    ManageIQ::Providers::Azure::CloudManager::Vm.find_by(:ems_ref => ems_ref)
+  end
+
+  def gather_storage_account_properties
+    sas = nil
+    source.with_provider_connection do |azure|
+      sas = Azure::Armrest::StorageAccountService.new(azure)
+    end
+    return if sas.nil?
+
+    begin
+      key             = sas.list_account_keys(storage_account_name, resource_group).fetch('key1')
+      storage_account = sas.get(storage_account_name, resource_group)
+      blob            = storage_account.blobs("system", key).first
+      blob_properties = storage_account.blob_properties(blob.container, blob.name, key)
+      endpoint        = storage_account.properties.primary_endpoints.blob
+      source_uri      = File.join(endpoint, blob.container, blob.name)
+      target_uri      = File.join(endpoint, "manageiq", dest_name + "_" + SecureRandom.uuid + ".vhd")
+    rescue Azure::Armrest::ResourceNotFoundException => err
+      _log.error("Error Class=#{err.class.name}, Message=#{err.message}")
+    end
+
+    return target_uri, source_uri, blob_properties.x_ms_meta_microsoftazurecompute_ostype
+  end
+
+  def prepare_for_clone_task
+    nic_id        = create_nic
+    root_password = MiqPassword.decrypt(options[:root_password]) if options[:root_password]
+    target_uri, source_uri, os = gather_storage_account_properties
+
+    {
+      :name       => dest_name,
+      :location   => source.location,
+      :properties => {
+        :hardwareProfile => {
+          :vmSize => instance_type.name
+        },
+        :osProfile       => {
+          :adminUserName => username,
+          :adminPassword => root_password,
+          :computerName  => dest_name
+        },
+        :storageProfile => {
+          :osDisk => {
+            :createOption => 'FromImage',
+            :caching      => 'ReadWrite',
+            :name         => dest_name + SecureRandom.uuid + '.vhd',
+            :osType       => os,
+            :image        => {:uri => source_uri},
+            :vhd          => {:uri => target_uri}
+          }
+        },
+        :networkProfile => {
+          :networkInterfaces => [{:id => nic_id}]
+        }
+      }
+    }
+  end
+
+  def log_clone_options(clone_options)
+    dumpObj(clone_options, "#{_log.prefix} Clone Options: ", $log, :info)
+    dumpObj(options, "#{_log.prefix} Prov Options:  ", $log, :info, :protected =>
+    {:path => workflow_class.encrypted_options_field_regs})
+  end
+
+  def region
+    source.location
+  end
+
+  def resource_group
+    source.description.split("\\").first
+  end
+
+  def username
```

Comment:
Do you need a whole method for this, or can it live inside `prepare_for_clone_task` with the other variables?